File: ria/java/gof/src/main/java/com/gof/pattern/factoryMethod/example/OrderProcessorFactory.java
package com.gof.pattern.factoryMethod.example;

public class OrderProcessorFactory {
    public static OrderProcessor createOrderProcessor(String orderType) {
        switch (orderType) {
            case "online":
                return new OnlineOrderProcessor();
            case "inStore":
                return new InStoreOrderProcessor();
            case "phone":
                return new PhoneOrderProcessor();
            default:
                throw new IllegalStateException("존재하지 않는 주문 타입니다. " + orderType);
        }
    }
}

File: ria/gof/src/main/kotlin/com/gof/pattern/example/OrderProcessorFactory.kt
package com.gof.pattern.example

class OrderProcessorFactory {
    companion object {
        fun createOrderProcessor(orderType: OrderType): OrderProcessor {
            return when (orderType) {
                OrderType.ONLINE ->
                    OnlineOrderProcessor()
                OrderType.OFFLINE ->
                    OfflineOrderProcessor()
                OrderType.PHONE ->
                    PhoneOrderProcessor()
                else ->
                    throw java.lang.IllegalArgumentException("존재하지 않는 주문 타입입니다. $orderType")
            }
        }
    }
}
